File: frontend/src/lib/schemas/skeleton.ts
// Mirrors backend/schemas/skeleton.py. Keep in sync.
import { z } from "zod";

export const checklistSourceSchema = z.enum([
  "jd_must_have",
  "jd_nice_to_have",
  "user_added",
]);
export const slotCoverageStatusSchema = z.enum([
  "uncovered",
  "mentioned",
  "demonstrated",
]);

export const skeletonInputSchema = z.object({
  user_id: z.string().optional(),
  job_description: z.string().min(10),
  job_url: z.string().optional(),
  target_seniority: z.string().optional(),
});

export const bulletPromptSchema = z.object({
  prompt: z.string().min(1),
  related_skills: z.array(z.string()).default([]),
});

export const experienceSlotSchema = z.object({
  slot_id: z.string().min(1),
  suggested_count_reason: z.string().optional(),
  company: z.string().default("[Company name]"),
  title: z.string().default("[Most recent role]"),
  dates: z.string().default("[Start date – End date]"),
  location: z.string().default("[City, Country or Remote]"),
  bullet_prompts: z.array(bulletPromptSchema).default([]),
  tech_stack_suggestions: z.array(z.string()).default([]),
});

export const projectSlotSchema = z.object({
  slot_id: z.string().min(1),
  name: z.string().default("[Project name]"),
  role: z.string().default("[Your role]"),
  bullet_prompts: z.array(bulletPromptSchema).default([]),
  tech_stack_suggestions: z.array(z.string()).default([]),
});

export const skillsChecklistItemSchema = z.object({
  name: z.string().min(1),
  source: checklistSourceSchema,
  checked: z.boolean().default(false),
  suggested_section: z.string().optional(),
});

export const coverageMapEntrySchema = z.object({
  status: slotCoverageStatusSchema.default("uncovered"),
  covered_by: z.string().nullable().optional(),
  where: z.string().nullable().optional(),
});

export const resumeSkeletonSchema = z.object({
  suggested_sections: z.array(z.string()).default([]),
  experience_slots: z.array(experienceSlotSchema).default([]),
  projects_slots: z.array(projectSlotSchema).default([]),
  skills_checklist: z.array(skillsChecklistItemSchema).default([]),
  coverage_map: z.record(z.string(), coverageMapEntrySchema).default({}),
  inferred_seniority: z.string().nullable().optional(),
});

export type SkeletonInput = z.infer<typeof skeletonInputSchema>;
export type BulletPrompt = z.infer<typeof bulletPromptSchema>;
export type ExperienceSlot = z.infer<typeof experienceSlotSchema>;
export type ProjectSlot = z.infer<typeof projectSlotSchema>;
export type SkillsChecklistItem = z.infer<typeof skillsChecklistItemSchema>;
export type CoverageMapEntry = z.infer<typeof coverageMapEntrySchema>;
export type ResumeSkeleton = z.infer<typeof resumeSkeletonSchema>;

File: backend/schemas/skeleton.py
# backend/schemas/skeleton.py
"""JD-aware resume skeleton: every career field is a placeholder string.

Used when the user has no base profile yet (or wants a structured starting
point). Field defaults are bracketed placeholders so an unfilled skeleton
trivially fails the anti-hallucination validator's "real OR placeholder"
check — nothing here pretends to be the user's actual history.
"""
from typing import Dict, List, Optional, Literal
from pydantic import BaseModel, Field


ChecklistSource = Literal["jd_must_have", "jd_nice_to_have", "user_added"]
SlotCoverageStatus = Literal["uncovered", "mentioned", "demonstrated"]


class SkeletonInput(BaseModel):
    user_id: Optional[str] = None
    job_description: str = Field(min_length=10)
    job_url: Optional[str] = None
    target_seniority: Optional[str] = None  # overrides JD-inferred seniority


class BulletPrompt(BaseModel):
    prompt: str = Field(min_length=1)
    related_skills: List[str] = []


class ExperienceSlot(BaseModel):
    slot_id: str = Field(min_length=1)
    suggested_count_reason: Optional[str] = None
    company: str = "[Company name]"
    title: str = "[Most recent role]"
    dates: str = "[Start date – End date]"
    location: str = "[City, Country or Remote]"
    bullet_prompts: List[BulletPrompt] = []
    tech_stack_suggestions: List[str] = []


class ProjectSlot(BaseModel):
    slot_id: str = Field(min_length=1)
    name: str = "[Project name]"
    role: str = "[Your role]"
    bullet_prompts: List[BulletPrompt] = []
    tech_stack_suggestions: List[str] = []


class SkillsChecklistItem(BaseModel):
    name: str = Field(min_length=1)
    source: ChecklistSource
    checked: bool = False
    suggested_section: Optional[str] = None  # "experience"|"projects"|"skills"


class CoverageMapEntry(BaseModel):
    status: SlotCoverageStatus = "uncovered"
    covered_by: Optional[str] = None  # slot_id of the demonstrating entry
    where: Optional[str] = None  # "experience.slot-1.bullets[2]" path


class ResumeSkeleton(BaseModel):
    suggested_sections: List[str] = []
    experience_slots: List[ExperienceSlot] = []
    projects_slots: List[ProjectSlot] = []
    skills_checklist: List[SkillsChecklistItem] = []
    coverage_map: Dict[str, CoverageMapEntry] = {}  # skill_name -> entry
    inferred_seniority: Optional[str] = None
